components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,218 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashSet;
+use std::str::Utf8Error;
+
+use adblock::engine::Engine as InnerEngine;
+use adblock::lists::FilterSet;
+use adblock::resources::{MimeType, Resource, ResourceType};
+use adblock::url_parser::ResolvesDomain;
+use cxx::{let_cxx_string, CxxString, CxxVector};
+
+use crate::ffi::{
+  resolve_domain_position, BlockerDebugInfo, BlockerResult, BlockerResultResult, BoolResult,
+  BoxEngineResult, EmptyTupleResult, FilterListMetadata, FilterListMetadataResult,
+  RegexManagerDiscardPolicy, StringResult, VecStringResult,
+};
+use crate::result::InternalError;
+
+pub struct Engine {
+  engine: InnerEngine,
+}
+
+impl Default for Box<Engine> {
+  fn default() -> Self {
+    new_engine()
+  }
+}
+
+pub fn new_engine() -> Box<Engine> {
+  Box::new(Engine {
+    engine: InnerEngine::new(true),
+  })
+}
+
+pub fn engine_with_rules(rules: &CxxVector<u8>) -> BoxEngineResult {
+  || -> Result<Box<Engine>, InternalError> {
+    let mut filter_set = FilterSet::new(false);
+    filter_set.add_filter_list(std::str::from_utf8(rules.as_slice())?, Default::default());
+    let engine = InnerEngine::from_filter_set(filter_set, true);
+    Ok(Box::new(Engine { engine }))
+  }()
```

Comment:
yeah, using a scope would be really nice. i tried to do this, but the `?` operator seems to only work if the surrounding function returns a Result that the error can be propagated to. Since a scope is not a function, i'm not sure how i can get this to work.

Reply:
Ah, I see. It'd be possible to implement [`std::ops::Try`](https://doc.rust-lang.org/std/ops/trait.Try.html) on these "fake" result types, which would then get it working - but that's unstable so let's leave it this way for now.